napari/components/viewer_model.py

```diff
@@ -354,16 +354,57 @@ def _on_active_layer(self, event):
     def rounded_division(min_val, max_val, precision):
         return int(((min_val + max_val) / 2) / precision) * precision
 
-    def _on_layers_change(self):
+    def _on_layers_change(self, *, add_layer=False):
+        new_point = None
         if len(self.layers) == 0:
             self.dims.ndim = 2
             self.dims.reset()
         else:
             ranges = self.layers._ranges
+            # If the point of an axis is outside of the upper current range,
+            # set it to the layers range max value. dims.set_current_step
+            # accounts for the clamping of the lower limit on the set_point call.
+            # New dimensions are prepended, hence we need to go through the list
+            # from the back to the beginning and invert the order in the end.
+            # To have the correct slices present.
+            prev_point = [
+                range_val[1] - 1 if point_val >= range_val[1] else point_val
+                for point_val, range_val in zip(
+                    reversed(self.dims.point), reversed(ranges)
+                )
+            ][::-1]
             ndim = len(ranges)
             self.dims.ndim = ndim
             self.dims.set_range(range(ndim), ranges)
 
+            # This logic block defines the behavior when layers are added or
+            # removed. Its goal is it to keep the currently selected view
+            # selected by the user.
+            if add_layer and len(self.layers) == 1:
```

Comment:
Not sure if I've fully thought this through, but what do you think about moving this logic inside `Dims.set_range`? I think the logic should be generally applicable whenever the ranges change. I think this is the only place (other than tests) that `set_range` is called, but that might not be true in the future. The other advantage is that this logic is pretty much entirely about `Dims`

Reply:
I moved the logic to the Dims class. I could not convert it fully into the `set_range` function though, because the point value changes on `self.dims.ndim = ndim`.
This triggers the `_check_dims` root_validator function of `Dims` which adjusts the range, i.e., the point property.
I tried to find a compromise :)

Reply:
Ah, good point on the root validator. That's unfortunate.

Reply:
Please add information about this to docstring.

Reply:
I was not 100% sure which docstring you meant, so I tried my best ^^